sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs

```diff
@@ -98,41 +105,115 @@ public class EventProcessor<T> where T : BasePartitionProcessor, new()
         private Dictionary<string, PartitionOwnership> InstanceOwnership { get; set; }
 
         /// <summary>
-        ///   The running task responsible for performing partition load balancing between multiple <see cref="EventProcessor{T}" />
+        ///   The running task responsible for performing partition load balancing between multiple <see cref="EventProcessor" />
         ///   instances, as well as managing partition pumps and ownership.
         /// </summary>
         ///
         private Task RunningTask { get; set; }
 
         /// <summary>
-        ///   Initializes a new instance of the <see cref="EventProcessor{T}"/> class.
+        ///   The function to be called just before event processing starts for a given partition.
         /// </summary>
         ///
-        /// <param name="consumerGroup">The name of the consumer group this event processor is associated with.  Events are read in the context of this group.</param>
-        /// <param name="eventHubClient">The client used to interact with the Azure Event Hubs service.</param>
-        /// <param name="partitionManager">Interacts with the storage system with responsibility for creation of checkpoints and for ownership claim.</param>
-        /// <param name="options">The set of options to use for this event processor.</param>
+        public Func<PartitionContext, Task> InitializeProcessingForPartitionAsync
+        {
+            internal get => _initializeProcessingForPartitionAsync;
+
+            set
+            {
+                lock (_setHandlerMutex)
+                {
+                    if (RunningTask == null)
+                    {
+                        _initializeProcessingForPartitionAsync = value;
+                    }
+                    else
+                    {
+                        throw new InvalidOperationException(Resources.RunningEventProcessorCannotPerformOperation);
+                    }
+                }
```

Comment:
@jsquire This pattern repeats itself four times in this code, on for each handler. Any ideas on how to avoid this repetition in a clean way?

Reply:
Since you're using the same lock primitive, you could wrap it in a helper method.    I'd also make sure you're doing a double-check lock pattern.

```csharp
if (RunningTask == null)
{
    lock (_setHandlerLock)
    {
          if (RunningTask == null)
          {
              _initializeProcessingForPartitionAsync = value;
              return;
          }
    }

    throw new InvalidOperationException();
}
```